url20b0f880a5_jgraph_legacy_jgraph5_tgz-pJ8-org_jgraph_example_HelloWorldJ8/src/com/jgraph/example/GraphEdX.java

```diff
@@ -0,0 +1,642 @@
+			}
+		};
+		url = getClass().getClassLoader().getResource(
+				"org/jgraph/example/resources/expand.gif");
+		expand.putValue(Action.SMALL_ICON, new ImageIcon(url));
+		expand.setEnabled(false);
+		toolbar.add(expand);
+
+		// ExpandAll
+		expandAll = new AbstractAction() {
+			public void actionPerformed(ActionEvent e) {
+				Object[] allCells = graph.getDescendants(graph.getRoots());
+				Map nested = new Hashtable();
+				for (int i=0; i < allCells.length; i++) {
+					Map attributeMap = new Hashtable();
+					GraphConstants.setForeground(attributeMap, Color.BLACK);
+					nested.put(allCells[i], attributeMap);
+				}
+				graph.getModel().edit(nested, null, null, null);
+			}
+		};
+		url = getClass().getClassLoader().getResource(
+				"org/jgraph/example/resources/expandAll.gif");
+		expandAll.putValue(Action.SMALL_ICON, new ImageIcon(url));
+		expandAll.setEnabled(false);
+		toolbar.add(expandAll);
+		return toolbar;
+	}
+
+	public void serializeGraph() {
+		int returnValue = JFileChooser.CANCEL_OPTION;
+		initFileChooser();
+		returnValue = fileChooser.showSaveDialog(graph);
+		if (returnValue == JFileChooser.APPROVE_OPTION) {
+			Container parent = graph.getParent();
+			BasicMarqueeHandler marquee = graph.getMarqueeHandler();
+			graph.setMarqueeHandler(null);
+			try {
+				// Serializes the graph by removing it from the component
+				// hierarchy and removing all listeners from it. The marquee
+				// handler, begin an inner class of GraphEd, is not marked
+				// serializable and will therefore not be stored. This must
+				// be taken into account when deserializing a graph.
+				uninstallListeners(graph);
+				parent.remove(graph);
+				ObjectOutputStream out = new ObjectOutputStream(
+						new BufferedOutputStream(new FileOutputStream(
+								fileChooser.getSelectedFile())));
+				out.writeObject(graph);
+				out.flush();
+				out.close();
+			} catch (Exception e) {
+				e.printStackTrace();
+				JOptionPane.showMessageDialog(graph, e.getMessage(), "Error",
+						JOptionPane.ERROR_MESSAGE);
+			} finally {
+				// Adds the component back into the component hierarchy
+				graph.setMarqueeHandler(marquee);
+				if (parent instanceof JViewport) {
+					JViewport viewPort = (JViewport) parent;
+					viewPort.setView(graph);
+				} else {
+					// Best effort...
+					parent.add(graph);
+				}
+				// And reinstalls the listener
+				installListeners(graph);
+			}
+		}
+	}
+
+	public void deserializeGraph() {
+		int returnValue = JFileChooser.CANCEL_OPTION;
+		initFileChooser();
+		returnValue = fileChooser.showOpenDialog(graph);
+		if (returnValue == JFileChooser.APPROVE_OPTION) {
+			Container parent = graph.getParent();
+			BasicMarqueeHandler marqueeHandler = graph.getMarqueeHandler();
+			try {
+				uninstallListeners(graph);
+				parent.remove(graph);
+				ObjectInputStream in = new ObjectInputStream(
+						new BufferedInputStream(new FileInputStream(fileChooser
+								.getSelectedFile())));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource created by the following constructor: *BufferedInputStream*. Currently, there are execution paths that do not contain closure statements, for example, when an if condition at line 359 evaluates to true. Ensure the object returned by *BufferedInputStream()* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).